### PR TITLE
Improve extended errors (rustc --explain).

### PR DESCRIPTION
**Pull request by michaelsproul**
I've taken another look at extended errors - fixing up the printing and adding a few more for match expressions.

With regards to printing, the previous behaviour was to just print the error message string directly, despite it containing indentation which caused it to overflow the standard terminal width of 80 columns (try `rustc --explain E0004`). The first approach I considered was to strip the leading whitespace from each line and lay out the text dynamically, inserting spaces in between. This approach became quite messy when taking multi-paragraph errors into account (and seemed overkill). The approach I settled on removes the indentation in the string itself and begins each message with a newline that is stripped before printing.

I feel like complete extended errors would be nice to have for 1.0.0 and I'm happy to spearhead an effort to get them written. Brian got me onto writing them at an SF meetup and I think it shouldn't be too hard to get the remaining 80 or so written with the help of people who don't really work on compiler innards.

### CODE REVIEW

File: src/librustc/diagnostics.rs [hunk 1]
```diff
@@ -10,51 +10,134 @@
...
+
+An example of an empty type is `enum Empty { }`.
 "##,
 
-    E0003: r##"
-    Not-a-Number (NaN) values can not be compared for equality and hence can never match
-    the input to a match expression. To match against NaN values, you should instead use
-    the `is_nan` method in a guard, as in: x if x.is_nan() => ...
+E0003: r##"
+Not-a-Number (NaN) values can not be compared for equality and hence can never
```
Comment (apasel422) [inline src/librustc/diagnostics.rs]: s/can not/cannot/

File: src/librustc/diagnostics.rs [hunk 2]
```diff
@@ -10,51 +10,134 @@
...
+match the input to a match expression. To match against NaN values, you should
+instead use the `is_nan` method in a guard, as in: x if x.is_nan() => ...
 "##,
 
-    E0004: r##"
-    This error indicates that the compiler can not guarantee a matching pattern for one
-    or more possible inputs to a match expression. Guaranteed matches are required in order
-    to assign values to match expressions, or alternatively, determine the flow of execution.
+E0004: r##"
+This error indicates that the compiler can not guarantee a matching pattern for
```
Comment (apasel422) [inline src/librustc/diagnostics.rs]: s/can not/cannot/

File: src/librustc/diagnostics.rs [hunk 3]
```diff
@@ -10,51 +10,134 @@
...
+E0006: r##"
+Patterns used to bind names must be irrefutable, that is, they must guarantee that a
+name will be extracted in all cases. If you encounter this error you probably need
+to use a `match` or `if let` to deal with the possibility of failure.
+"##,
+
+E0007: r##"
+This error indicates that the bindings in a match arm would require a value to
+be moved into more than one location, thus violating unique ownership. Code like
+the following is invalid as it requires the entire Option<String> to be moved
```
Comment (apasel422) [inline src/librustc/diagnostics.rs]: Option\<String\> should be surrounded by ` for consistency.

**Comment by pnkfelix** [comment]
@michaelsproul Is this commit just changing the code formatting and where the line breaks fall?  (Or at least intended to solely to change that?)

Or are there changing to the content mixed in?

**Comment by michaelsproul** [comment]
@pnkfelix: Yeah 19e9828 doesn't change the content at all, it just reflows it in some places.

**Comment by pnkfelix** [comment]
r=me after the example in ad2e506 is fixed in some manner.

**Comment by michaelsproul** [comment]
Fixed the `&Option<u32>` thing and adopted the "cannot"s, although I managed to lose your comment in the process of force pushing (whoops).

Thanks for reviewing :smile:

**Comment by pnkfelix** [comment]
@michaelsproul the point from my earlier comment was that i was not sure whether the provided suggested revision was quite right ... although now I see that you have addressed the most important point (that `op_num` changed its type in the revision), so I guess you probably did see the original comment.

So, okay.

**Comment by michaelsproul** [comment]
@pnkfelix: Yeah, I think the use of a `Copy` type makes the `op_num` example less clear than it could be. Perhaps something like this would serve us better:

```rust
// Before.
match Some("hi".to_string()) {
    ref op_string_ref @ Some(ref s) => ...
    None => ...
}

// After.
match Some("hi".to_string()) {
    Some(ref s) => {
        let opt_string_ref = &Some(&s);
        ...
    }
    None => ...
}
```

This makes a bit more sense, with `op_string_ref : &Option<&String>`.

If you like this change I could push it and we could approve the new commit before bors jumps on the old one (or it could wait for the next PR).

**Comment by brson** [comment]
Awesome! This is a really important polish area for 1.0. Anything you can do to stir up more support for this effort would be great.

**Pull request merged**